### PR TITLE
feat: Experimentally introducing sink into table

### PR DESCRIPTION
**Pull request by shanicky**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR depends on #12240 and #12495

This PR attempts to introduce the sink into table feature.

Following PR #12240, all newly created tables come with a union operator used to aggregate source/dml and the current sink into table. According to the rfc, we create various structures of sink under these three types.

The specific method is very rudimentary. Currently, when we create a streaming job, we will have an `affected_table_change`, which describes the target table that needs to replace the table. The generated command carries both add and update semantics at this time. We tried to use the combined mutation in PR #12495 to perform this process atomically, but it seems that it may not be necessary, so we may delete this mutation after testing.

Mainly, we still have many restrictions that need to be solved in the future:

1. We can only create 1 sink to the table at present.
2. Tables with sinks cannot be altered at the moment.

### The first case, table with row id as the primary key.
image

```
dev=> create table m (v1 int, v2 int, v3 int as v1 + v2)  with (
    connector = 'datagen',
    fields.v1.kind = 'sequence',
    fields.v1.start = '1',
    fields.v1.end = '1000',
    datagen.split.num = '5',
) FORMAT PLAIN ENCODE JSON ;
CREATE_TABLE
dev=> create table t (v1 int);
CREATE_TABLE
dev=> explain create sink s into m as select v1 from t with ( type = 'append-only' , force_append_only = 'true');
                                       QUERY PLAN
-----------------------------------------------------------------------------------------
 StreamProject { exprs: [t.v1, null:Int32, (t.v1 + null:Int32) as $expr1, null:Serial] }
 └─StreamProject { exprs: [t.v1, null:Int32, null:Serial] }
   └─StreamSink { type: append-only, columns: [v1, t._row_id(hidden)] }
     └─StreamTableScan { table: t, columns: [v1, _row_id] }
(4 rows)
dev=> create sink s into m as select v1 from t with ( type = 'append-only' , force_append_only = 'true');
CREATE_SINK
```

#### table
image

#### sink
image

### The second case, append only table

image

```
dev=> create table m (v1 int, v2 int, v3 int as v1 + v2) append only with (
    connector = 'datagen',
    fields.v1.kind = 'sequence',
    fields.v1.start = '1',
    fields.v1.end = '1000',
    datagen.split.num = '5',
) FORMAT PLAIN ENCODE JSON ;
NOTICE:  APPEND ONLY TABLE is currently an experimental feature.
CREATE_TABLE
dev=> create table t (v1 int);
CREATE_TABLE
dev=> explain create sink s into m as select v1 from t with ( type = 'append-only' , force_append_only = 'true');
                                       QUERY PLAN
-----------------------------------------------------------------------------------------
 StreamProject { exprs: [t.v1, null:Int32, (t.v1 + null:Int32) as $expr1, null:Serial] }
 └─StreamProject { exprs: [t.v1, null:Int32, null:Serial] }
   └─StreamSink { type: append-only, columns: [v1, t._row_id(hidden)] }
     └─StreamTableScan { table: t, columns: [v1, _row_id] }
(4 rows)
dev=> create sink s into m as select v1 from t with ( type = 'append-only' , force_append_only = 'true');
CREATE_SINK
```

#### table
image

#### sink
image

### The third case,table with primary key

image

```
dev=> create table m (v1 int primary key, v2 int) with {
    connector = 'datagen',
    fields.v1.kind = 'sequence',
    fields.v1.start = '1',
    fields.v1.end = '1000',
    datagen.split.num = '5',
) FORMAT PLAIN ENCODE JSON ;
CREATE_TABLE
dev=> create table t (v1 int);
CREATE_TABLE
dev=> explain create sink s into m as select v1 from t;
                                    QUERY PLAN
----------------------------------------------------------------------------------
 StreamProject { exprs: [t.v1, null:Int32] }
 └─StreamSink { type: upsert, columns: [v1, t._row_id(hidden)], pk: [t._row_id] }
   └─StreamTableScan { table: t, columns: [v1, _row_id] }
(3 rows)
dev=> create sink s into m as select v1 from t;
CREATE_SINK
```

#### table
image

#### sink
image

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/meta/src/rpc/ddl_controller.rs
Comment (yezizp2012) [inline src/meta/src/rpc/ddl_controller.rs:579]: Please add some strdoc for it.

Comment (yezizp2012) [reply src/meta/src/rpc/ddl_controller.rs:579]: Can we change this logic as a last part of build streaming job if replace info provided?

File: src/meta/src/rpc/ddl_controller.rs
Comment (yezizp2012) [inline src/meta/src/rpc/ddl_controller.rs:599]: Please add some comments for this function.

File: src/meta/src/rpc/ddl_controller.rs
Comment (yezizp2012) [inline src/meta/src/rpc/ddl_controller.rs:681]:
```suggestion
```
Duplicated.

File: src/meta/src/rpc/ddl_controller.rs
Comment (yezizp2012) [inline src/meta/src/rpc/ddl_controller.rs:591]: Please add some assertion for some frontend work.

File: src/meta/src/stream/stream_graph/fragment.rs
Comment (yezizp2012) [inline src/meta/src/stream/stream_graph/fragment.rs:508]: Remove it since it's never used.

File: src/meta/src/stream/stream_manager.rs
Comment (yezizp2012) [inline src/meta/src/stream/stream_manager.rs:476]: Still need to check create type for it. FYI, we don't support create sink in background for now, but we still need take this feature into consideration after it get supported.

File: src/meta/src/stream/stream_manager.rs
Comment (yezizp2012) [inline src/meta/src/stream/stream_manager.rs:503]: Can we unify some logic for this part with replace table somehow?

File: src/meta/service/src/ddl_service.rs
Comment (yezizp2012) [inline src/meta/service/src/ddl_service.rs:321]: Unify this part with `create`, they look exactly the same.

File: src/meta/src/barrier/command.rs
Comment (yezizp2012) [inline src/meta/src/barrier/command.rs:76]:
```suggestion
pub struct ReplaceTablePlan {
```
Unify it with `Command::ReplaceTable`, and better not name it as command, but I don't have any better naming. 🥵

File: src/meta/src/manager/catalog/mod.rs
Comment (yezizp2012) [inline src/meta/src/manager/catalog/mod.rs:2366]: Did you meet any failure here while doing any local test? Better to resolve it.

Comment (yezizp2012) [reply src/meta/src/manager/catalog/mod.rs:2366]: Oh, we need to check the table version as well for create sink into table, since you need to replace the whole table streaming job as well like `alter table`.

File: src/frontend/src/handler/create_table.rs
Comment (yezizp2012) [inline src/frontend/src/handler/create_table.rs:1116]: Should fill the table version in frontend, which should be checked in meta side to do table replacement.

**Review by yezizp2012 — COMMENTED** [review]
Rest of meta part generally LGTM.

**Comment by BugenZhao** [comment]
Will review this PR after it get rebased. 🫡

File: src/meta/src/rpc/ddl_controller.rs
Comment (fuyufjh) [inline src/meta/src/rpc/ddl_controller.rs:1029]: The struct `replace_table_info` was passed via a deep code path. I am thinking that is it possible to simplify it like this?

```
RPC: create_sink -> run_command -> create_streaming_job -> build_stream_job -> inject_replace_table_job
⬇️
RPC: create_sink -> run_command (1) original code path
                 -> run_command (2) inject_replace_table_job
```

(or can be even better if they can be seperated into multiple RPCs but with atomicity... but this seems to be more difficult)

File: src/stream/src/executor/dispatch.rs
Comment (fuyufjh) [inline src/stream/src/executor/dispatch.rs:309]: Why not just change the mutation to a list of mutations in proto?

```diff
 message Barrier {
-  BarrierMutation mutation = 3;
+  repeated BarrierMutation mutations = 3;
```

**Review by fuyufjh — COMMENTED** [review]
Generally LGTM

File: src/frontend/src/handler/create_sink.rs
Comment (st1page) [inline src/frontend/src/handler/create_sink.rs:227]:
```suggestion
                unreachable!("can not derive generated columns in a sink's catalog, but meet one");
```

File: src/frontend/src/handler/create_sink.rs
Comment (st1page) [inline src/frontend/src/handler/create_sink.rs:260]: We might need to check it on the meta because https://github.com/risingwavelabs/risingwave/issues/5895?
or could it be rejected by other mechanisms?

File: e2e_test/sink/sink_into_table.slt
Comment (st1page) [inline e2e_test/sink/sink_into_table.slt:26]: suggest add test to drop table before drop sink and see if there is an error.

File: src/frontend/src/handler/create_sink.rs
Comment (st1page) [inline src/frontend/src/handler/create_sink.rs:473]:
```suggestion
pub(crate) fn insert_merger_to_union(node: &mut StreamNode, identity: &str) {
```
so that we can leave the table sink's id and name here?

File: src/meta/src/controller/mod.rs
Comment (yezizp2012) [inline src/meta/src/controller/mod.rs:138]:
```suggestion
	   // TODO: fix it for model v2.
            incoming_sinks: vec![],
```

**Review by yezizp2012 — COMMENTED** [review]
Generally LGTM, may wait for review from @BugenZhao .

File: src/stream/src/executor/dispatch.rs
Comment (fuyufjh) [reply src/stream/src/executor/dispatch.rs:307]: Any updates?

Comment (fuyufjh) [reply src/stream/src/executor/dispatch.rs:307]: Any updates here? 👀

File: src/frontend/src/handler/create_sink.rs
Comment (BugenZhao) [inline src/frontend/src/handler/create_sink.rs:320]: Looks like the `definition` is never mutated.

File: src/frontend/src/handler/create_sink.rs
Comment (BugenZhao) [inline src/frontend/src/handler/create_sink.rs:375]: Is this step necessary? As the `columns` are not changed, the mapping should always be an identical mapping. Can we simply fill `None` and let the meta service skip this?

File: src/frontend/src/handler/create_table.rs
Comment (BugenZhao) [reply src/frontend/src/handler/create_table.rs:1150]: Since the `col_id_gen` passed in is `new_alter`, I guess the `Table` returned by `gen_create_table_plan` has already bumped its version.

File: src/frontend/src/handler/drop_sink.rs
Comment (BugenZhao) [inline src/frontend/src/handler/drop_sink.rs:66]: Can we extract the common codes with `create_sink` into a function?

File: src/stream/src/executor/dispatch.rs
Comment (BugenZhao) [reply src/stream/src/executor/dispatch.rs:307]: However, is it general enough for all executors to correctly handle it?

File: src/stream/src/executor/dispatch.rs
Comment (BugenZhao) [inline src/stream/src/executor/dispatch.rs:238]: As there's only a single form of combination here, what about directly naming it as `AddAndUpdate`? 🤣 Then it can also be type-safer.

This requires us to extract the fields of `Mutation::Add` and `Mutation::Update` to dedicated structs.

File: src/frontend/src/handler/create_sink.rs
Comment (BugenZhao) [reply src/frontend/src/handler/create_sink.rs]: Oh, I find it already been changed to `Option`. Cool!

Comment (BugenZhao) [reply src/frontend/src/handler/create_sink.rs]: Would you mind document it in `.proto` for when it's `None`?

File: src/meta/src/barrier/command.rs
Comment (BugenZhao) [inline src/meta/src/barrier/command.rs:187]:
```suggestion
                replace_table: Some(ReplaceTablePlan {
```

File: src/meta/src/manager/catalog/mod.rs
Comment (BugenZhao) [inline src/meta/src/manager/catalog/mod.rs:1458]: Does it mean that cascadely dropping a view with table sink dependencies will be rejected?

File: src/sqlparser/src/ast/statement.rs
Comment (BugenZhao) [inline src/sqlparser/src/ast/statement.rs:853]: Do we expect non-empty with properties for `SINK INTO TABLE`?

File: src/meta/src/rpc/ddl_controller.rs
Comment (BugenZhao) [inline src/meta/src/rpc/ddl_controller.rs:636]: Can we do this in the frontend to make meta's code clearer?

File: src/meta/src/rpc/ddl_controller.rs
Comment (BugenZhao) [inline src/meta/src/rpc/ddl_controller.rs:880]: Will these two operations be done atomically?

Comment (BugenZhao) [reply src/meta/src/rpc/ddl_controller.rs:880]: The `cancel_stream_job` above is fallible, which looks not good. 😕

File: src/meta/src/stream/stream_manager.rs
Comment (BugenZhao) [inline src/meta/src/stream/stream_manager.rs:568]: Same concern here.

File: e2e_test/sink/sink_into_table.slt
Comment (BugenZhao) [inline e2e_test/sink/sink_into_table.slt:1]: Would you mind adding planner tests for `SINK INTO TABLE`? Maybe in next PRs.

**Review by BugenZhao — COMMENTED** [review]
Rest generally LGTM. Great job!

File: src/meta/src/rpc/ddl_controller.rs
Comment (st1page) [reply src/meta/src/rpc/ddl_controller.rs:636]: Yes, because the metadata on frontend could be stale.

Comment (BugenZhao) [reply src/meta/src/rpc/ddl_controller.rs:636]: That's not a big deal. 😄 We have a version number in the table catalog. If the catalog is stale, it'll be rejected. In other words, there're still race cases if two DDLs are running in the meta service.

File: src/meta/src/rpc/ddl_controller.rs
Comment (BugenZhao) [reply src/meta/src/rpc/ddl_controller.rs]: Does using a different error message work here?

File: src/meta/src/rpc/ddl_controller.rs
Comment (BugenZhao) [reply src/meta/src/rpc/ddl_controller.rs:627]: Okay, let's see if there is a better way to construct the codes once all the functionalities are completed.

**Review by BugenZhao — APPROVED** [review]
LGTM. Really a masterpiece!

**Review by yezizp2012 — APPROVED** [review]
Great job!LGTM!

**Pull request merged**
